src/server/server_family.cc

```diff
@@ -2025,10 +2040,111 @@ void ServerFamily::Register(CommandRegistry* registry) {
+      << CI{"SLOWLOG", CO::ADMIN | CO::FAST, -2, 0, 0, 0, acl::kSlowLog}.HFUNC(SlowLog)
       << CI{"SCRIPT", CO::NOSCRIPT | CO::NO_KEY_TRANSACTIONAL, -2, 0, 0, 0, acl::kScript}.HFUNC(
              Script)
       << CI{"DFLY", CO::ADMIN | CO::GLOBAL_TRANS | CO::HIDDEN, -2, 0, 0, 0, acl::kDfly}.HFUNC(Dfly);
 }
 
+void ServerFamily::SlowLog(CmdArgList args, ConnectionContext* cntx) {
+  ToUpper(&args[0]);
+  string_view sub_cmd = ArgS(args, 0);
+
+  if (sub_cmd == "HELP") {
+    string_view help[] = {
+        "SLOWLOG <subcommand> [<arg> [value] [opt] ...]. Subcommands are:",
+        "GET [<count>]",
+        "    Return top <count> entries from the slowlog (default: 10, -1 mean all).",
+        "    Entries are made of:",
+        "    id, timestamp, time in microseconds, arguments array, client IP and port,",
+        "    client name",
+        "LEN",
+        "    Return the length of the slowlog.",
+        "RESET",
+        "    Reset the slowlog.",
+        "HELP",
+        "    Prints this help.",
+    };
+    (*cntx)->SendSimpleStrArr(help);
+    return;
+  }
+
+  if (sub_cmd == "LEN") {
+    size_t sum = 0;
+    service_.proactor_pool().AwaitFiberOnAll(
+        [&sum](auto index, auto* context) { sum += ServerState::tlocal()->GetSlowLog().Length(); });
+    return (*cntx)->SendLong(sum);
+  }
+
+  if (sub_cmd == "RESET") {
+    service_.proactor_pool().AwaitFiberOnAll(
+        [](auto index, auto* context) { ServerState::tlocal()->GetSlowLog().Reset(); });
+    return (*cntx)->SendOk();
+  }
+
+  if (sub_cmd == "GET") {
+    size_t requested_slow_log_length = UINT32_MAX;
+
+    if (args.size() == 2) {
+      string_view length = ArgS(args, 1);
+      int64_t num;
+      if ((!absl::SimpleAtoi(length, &num)) || (num < -1)) {
+        (*cntx)->SendError("count should be greater than or equal to -1");
+        return;
+      }
+      if (num >= 0) {
+        requested_slow_log_length = num;
+      }
+    }
+
+    // gather all the individual slowlogs from all the fibers and sort them by their timestamp
+    std::vector<std::pair<SlowLogEntry, unsigned>> entries;
+    service_.proactor_pool().AwaitFiberOnAll([&entries](auto index, auto* context) {
+      for (const auto& logItem : ServerState::tlocal()->GetSlowLog().Entries()) {
+        entries.emplace_back(logItem, index);
+      }
+    });
+
+    std::sort(entries.begin(), entries.end(), [](const auto& e1, const auto& e2) {
+      return e1.first.UnixTimestamp() > e2.first.UnixTimestamp();
+    });
+
+    requested_slow_log_length = std::min(entries.size(), requested_slow_log_length);
+
+    (*cntx)->StartArray(requested_slow_log_length);
+    for (size_t i = 0; i < requested_slow_log_length; ++i) {
+      const auto& entry = entries[i].first;
+      const auto& args = entry.CmdArgs();
+
+      (*cntx)->StartArray(6);
+
+      (*cntx)->SendLong(entry.CommandId() * service_.proactor_pool().size() + entries[i].second);
```

Comment:
This returned value should be a unique progressive identifier for the log entry. You mul the command id with number of threads and add the shard index to it, correct? if so , this is not a  unique progressive identifier

Reply:
@adiholden commandid is incremented in each thread. why do you think that `commandid*nthreads+threadid` is not a good mapping?

Reply:
This comment can be ignored. I thought command id is the command id from command registry, I made a comment to rename CommandId later